File: app/src/main/java/ru/mammoth70/wherearetheynow/DBhelper.java
package ru.mammoth70.wherearetheynow;

import android.content.Context;
import android.database.Cursor;
import android.database.SQLException;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;

import java.util.Locale;

public class DBhelper extends SQLiteOpenHelper {
    // Класс обслуживает базу данных со списком контактов и прочими структурами контактов.
    private static final int DB_VERSION = 3; // версия БД

    public DBhelper(Context context) {
        super(context, "watnDB", null, DB_VERSION);
    }

    private final String createTableUsersSting = "CREATE TABLE IF NOT EXISTS users " +
            "(id integer PRIMARY KEY AUTOINCREMENT, phone text UNIQUE, name text, color text);";
    private final String createTablePointsSting = "CREATE TABLE IF NOT EXISTS points " +
            "(phone text UNIQUE, latitude text, longitude text, datetime text);";

    @Override
    public void onCreate(SQLiteDatabase db) {
        // Метод создаёт таблицы users и points.
        db.execSQL(createTableUsersSting);
        db.execSQL(createTablePointsSting);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        // Метод делает апгрейд БД.
        if (oldVersion == 1 && newVersion > 1) {
            // Удаляем из таблицы users поле label.
            // Снимаем ограничение на уникальность поля name в таблице users.
            // Добавляем таблицу points.
            db.beginTransaction();
            try {
                db.execSQL("CREATE TEMPORARY TABLE users_temp " +
                        "(id integer, phone text, name text, color text);");
                db.execSQL("INSERT INTO users_temp SELECT id, phone, name, color FROM users;");
                db.execSQL("DROP TABLE users;");
                db.execSQL(createTableUsersSting);
                db.execSQL("INSERT INTO users SELECT id, phone, name, color FROM users_temp;");
                db.execSQL("DROP TABLE users_temp;");
                db.execSQL(createTablePointsSting);
                db.setTransactionSuccessful();
            } finally {
                db.endTransaction();
            }
        }
        if (oldVersion == 2 && newVersion > 2) {
            // Добавляем таблицу points.
            db.beginTransaction();
            try {
                db.execSQL(createTablePointsSting);
                db.setTransactionSuccessful();
            } finally {
                db.endTransaction();
            }
        }
    }

    public void getUsers() {
        // Метод считывает список разрешенных телефонов и словари контактов из БД.
        try (SQLiteDatabase db = getReadableDatabase()) {
            try (Cursor cursor = db.rawQuery("SELECT * FROM users;", null)) {
                Util.phones.clear();
                Util.id2phone.clear();
                Util.phone2id.clear();
                Util.phone2name.clear();
                Util.phone2color.clear();
                Util.phone2record.clear();
                while (cursor.moveToNext()) {
                    int id = cursor.getInt(cursor.getColumnIndexOrThrow("id"));
                    String phone = cursor.getString(cursor.getColumnIndexOrThrow("phone"));
                    String name = cursor.getString(cursor.getColumnIndexOrThrow("name"));
                    String color = cursor.getString(cursor.getColumnIndexOrThrow("color"));
                    Util.phones.add(phone);
                    Util.id2phone.put(id, phone);
                    Util.phone2id.put(phone, id);
                    Util.phone2name.put(phone, name);
                    Util.phone2color.put(phone, color);
                    PointRecord record = getLastPoint(phone);
                    if (record != null) {
                        Util.phone2record.put(phone, record);
                    }
                }
            }
        }
    }

    public void getMenuUsers() {
        // Метод считывает список телефонов, отортированный по id ASC,
        // имеющих координаты, отсортированные по дате DESC, и по id ASC,
        // и ограниченные заданным (10-ю) количеством записей.
        // Используется для построения меню.
        Util.menuPhones.clear();
        try (SQLiteDatabase db = getReadableDatabase()) {
            String execSting =
                    "SELECT users.phone AS phone1 FROM " +
                            "(SELECT points.phone AS phone2 " +
                            "FROM points INNER JOIN users ON points.phone = users.phone " +
                            "ORDER BY points.datetime DESC, users.id ASC LIMIT 10 OFFSET 0) AS tmp " +
                            "INNER JOIN users ON tmp.phone2 = users.phone ORDER BY users.id;";
            try (Cursor cursor = db.rawQuery(execSting, null)) {
                while (cursor.moveToNext()) {
                    String phone = cursor.getString(cursor.getColumnIndexOrThrow("phone1"));
                    Util.menuPhones.add(phone);
                }
            }
        }
    }

    public boolean addUser(String phone, String name, String color) {
        // Метод добавляет запись контакта в БД и обновляет структуры.
        // Возвращает true, если успешно и false, если нет.
        if (!Util.phones.contains(phone)) {
            try (SQLiteDatabase db = getReadableDatabase()) {
                String execSting = "INSERT OR IGNORE INTO users (phone, name, color) VALUES " +
                        "('" + phone + "', '" + name + "', '" + color + "');";
                try {
                    db.execSQL(execSting);
                } catch (SQLException ignored) {
                    return false;
                }
            }
            getUsers();
            return (Util.phones.contains(phone));
        } else {
            return false;
        }
    }

    public boolean editUser(int id, String phone, String name, String color) {
        // Метод изменяет запись контакта в БД и обновляет структуры.
        // Возвращает true, если успешно и false, если нет.
        if (Util.id2phone.containsKey(id)) {
            try (SQLiteDatabase db = getReadableDatabase()) {
                String execSting = "UPDATE users SET " +
                        "phone = '" + phone + "', " +
                        "name = '" + name + "', " +
                        "color = '" + color + "' " +
                        " WHERE id = '" + id + "';";
                try {
                    db.execSQL(execSting);
                } catch (SQLException ignored) {
                    return false;
                }
            }
            getUsers();
            return (Util.phones.contains(phone));
        } else {
            return false;
        }
    }

    public boolean deleteUser(int id) {
        // Метод удаляет запись контакта в БД и обновляет структуры.
        // Возвращает true, если успешно и false, если нет.
        if (Util.id2phone.containsKey(id)) {
            String phone = Util.id2phone.get(id);
            try (SQLiteDatabase db = getReadableDatabase()) {
                String execSting1 = "DELETE FROM users WHERE id = '" + id + "';";
                String execSting2 = "DELETE FROM points WHERE phone = '" + phone + "';";
                try {
                    db.execSQL(execSting1);
                    db.execSQL(execSting2);
                } catch (SQLException ignored) {
                    return false;
                }
            }
            getUsers();
            return (!Util.id2phone.containsKey(id));
        } else {
            return false;
        }
    }

    public void setLastPoint(PointRecord record) {
        // Метод заносит в таблицу points последние известные координаты контакта.
        if (Util.phones.contains(record.phone)) {
            String latitude = String.format(Locale.US, PointRecord.FORMAT_DOUBLE, record.latitude);
            String longitude = String.format(Locale.US, PointRecord.FORMAT_DOUBLE, record.longitude);
            if ((record.latitude > -90) && (record.latitude < 90) &&
                    (record.longitude > -180) && (record.longitude < 180)) {
                try (SQLiteDatabase db = getReadableDatabase()) {
                    String execSting = "INSERT OR REPLACE INTO points (phone, latitude, longitude, datetime)" +
                            " VALUES ('" + record.phone + "', '" +
                            latitude + "', '" + longitude + "', '" +
                            record.datetime + "');";
                    try {
                        db.execSQL(execSting);
                    } catch (SQLException ignored) {
                    }
                }
            }
        }
    }

    public PointRecord getLastPoint(String phone) {
        // Метод возвращает PointRecord по заданному телефону, или null, если нет или неправильно.
        if (Util.phones.contains(phone)) {
            try (SQLiteDatabase db = getReadableDatabase()) {
                String execSting = "SELECT * FROM points WHERE phone = '" + phone + "';";
                try (Cursor cursor = db.rawQuery(execSting, null)) {
                    if (cursor.moveToFirst()) {
                        PointRecord record = new PointRecord(
                                cursor.getString(cursor.getColumnIndexOrThrow("phone")),
                                Double.parseDouble(
                                        cursor.getString(cursor.getColumnIndexOrThrow("latitude"))),
                                Double.parseDouble(
                                        cursor.getString(cursor.getColumnIndexOrThrow("longitude"))),
                                cursor.getString(cursor.getColumnIndexOrThrow("datetime")));
                        if ((record.latitude > -90) && (record.latitude < 90) &&
                                (record.longitude > -180) && (record.longitude < 180)) {
                            return record;
                        } else {
                            return null;
                        }
                    }
                }
            }
        }
        return null;
    }

}

File: app/src/main/java/ru/mammoth70/wherearetheynow/DBhelper.kt
package ru.mammoth70.wherearetheynow

import android.content.ContentValues
import android.content.Context
import android.database.SQLException
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper
import androidx.core.database.sqlite.transaction
import ru.mammoth70.wherearetheynow.App.Companion.appContext
import java.util.Locale

class DBhelper(context: Context?) : SQLiteOpenHelper(context, "watnDB",
    null, DB_VERSION) {
    // Класс обслуживает базу данных со списком контактов и прочими структурами контактов.

    companion object {
        private const val DB_VERSION = 3 // версия БД
        val dbHelper = DBhelper(appContext)
        private const val CREATE_TABLE_USERS = "CREATE TABLE IF NOT EXISTS users " +
                "(id integer PRIMARY KEY AUTOINCREMENT, phone text UNIQUE, name text, color text);"
        private const val CREATE_TABLE_POINTS = "CREATE TABLE IF NOT EXISTS points " +
                "(phone text UNIQUE, latitude text, longitude text, datetime text);"
    }

    override fun onCreate(db: SQLiteDatabase) {
        // Функция создаёт таблицы users и points.
        db.execSQL(CREATE_TABLE_USERS)
        db.execSQL(CREATE_TABLE_POINTS)
    }

    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        // Функция делает апгрейд БД.

        if (oldVersion == 1 && newVersion > 1) {
            // Удаление из таблицы users поля label.
            // Удаление ограничения на уникальность поля name в таблице users.
            // Добавление таблицы points.
            db.transaction {
                try {
                    execSQL("CREATE TEMPORARY TABLE users_temp (id integer, phone text, name text, color text);")
                    execSQL("INSERT INTO users_temp SELECT id, phone, name, color FROM users;")
                    execSQL("DROP TABLE users;")
                    execSQL(CREATE_TABLE_USERS)
                    execSQL("INSERT INTO users SELECT id, phone, name, color FROM users_temp;")
                    execSQL("DROP TABLE users_temp;")
                    execSQL(CREATE_TABLE_POINTS)
                } finally {
                }
            }
        }
        if (oldVersion == 2 && newVersion > 2) {
            // Добавление таблицы points.
            db.transaction {
                try {
                    execSQL(CREATE_TABLE_POINTS)
                } finally {
                }
            }
        }
    }

    fun readDbAllUsers(): List<User> {
        // Функция считывает из БД данные контактов и координат.

        val userList = mutableListOf<User>()
        val query = """
        SELECT u.id, u.phone, u.name, u.color, p.latitude, p.longitude, p.datetime 
        FROM users u 
        LEFT JOIN points p ON u.phone = p.phone
    """.trimIndent()

        try {
            readableDatabase.rawQuery(query, null).use { cursor ->
                val idIdx = cursor.getColumnIndexOrThrow("id")
                val phoneIdx = cursor.getColumnIndexOrThrow("phone")
                val nameIdx = cursor.getColumnIndexOrThrow("name")
                val colorIdx = cursor.getColumnIndexOrThrow("color")
                val latIdx = cursor.getColumnIndexOrThrow("latitude")
                val lonIdx = cursor.getColumnIndexOrThrow("longitude")
                val dateIdx = cursor.getColumnIndexOrThrow("datetime")

                while (cursor.moveToNext()) {
                    val phone = cursor.getString(phoneIdx) ?: continue
                    val rawLat = cursor.getString(latIdx)
                    val rawLon = cursor.getString(lonIdx)
                    val dateTime = cursor.getString(dateIdx)

                    var lastRecord: PointRecord? = null

                    if (rawLat != null && rawLon != null && dateTime != null) {
                        val lat = rawLat.toDouble()
                        val lon = rawLon.toDouble()

                        if (lat in -90.0..90.0 && lon in -180.0..180.0) {
                            lastRecord = PointRecord(phone, lat, lon, dateTime)
                        }
                    }

                    val user = User(
                        id = cursor.getInt(idIdx),
                        phone = phone,
                        name = cursor.getString(nameIdx) ?: "",
                        color = cursor.getString(colorIdx) ?: "",
                        lastRecord = lastRecord
                    )
                    userList.add(user)
                }
            }
        } catch (e: NumberFormatException) {
            LogSmart.e("DBhelper", "NumberFormatException в readDbAllUsers()", e)
        } catch (e: SQLException) {
            LogSmart.e("DBhelper", "SQLException в readDbAllUsers()", e)
        }
        return userList
    }

    fun readDbMenuUsers(): List<String> {
        // Функция возвращает список телефонов, отсортированный по id ASC,
        // которые имеют координаты, отсортированные по дате DESC, и по id ASC,
        // и ограничены заданным (10-ю) количеством записей.
        // Функция используется для построения меню.

        val list = mutableListOf<String>()
        val query = """
            SELECT users.phone AS phone_result FROM
                (SELECT points.phone AS phone2
                FROM points INNER JOIN users ON points.phone = users.phone
                ORDER BY points.datetime DESC, users.id ASC LIMIT 10 OFFSET 0) AS tmp
                INNER JOIN users ON tmp.phone2 = users.phone ORDER BY users.id;
            """.trimIndent()
        try {
            readableDatabase.rawQuery(query, null).use { cursor ->
                val phoneIdx = cursor.getColumnIndexOrThrow("phone_result")
                while (cursor.moveToNext()) {
                    cursor.getString(phoneIdx)?.let { list.add(it) }
                }
            }
        } catch (e: SQLException) {
            LogSmart.e("DBhelper", "SQLException в readDbMenuUsers()", e)
        }
        return list
    }

    fun addDbUser(user: User): Long {
        // Функция добавляет контакт в БД.

        val values = ContentValues().apply {
            put("phone", user.phone)
            put("name", user.name)
            put("color", user.color)
        }

        return try {
            writableDatabase.insertWithOnConflict(
                "users",
                null,
                values,
                SQLiteDatabase.CONFLICT_IGNORE
            )
        } catch (e: SQLException) {
            LogSmart.e("DBhelper", "SQLException в addDbUser($user)", e)
            -1L
        }
    }

    fun editDbUser(user: User): Boolean {
        // Функция меняет контакт в БД.

        val values = ContentValues().apply {
            put("phone", user.phone)
            put("name", user.name)
            put("color", user.color)
        }

        return try {
            val rowsAffected = writableDatabase.update(
                "users",
                values,
                "id = ?",
                arrayOf(user.id.toString())
            )

            if (rowsAffected > 0) {
                checkDbRecords() // Чистим "висящие" точки в БД
                true
            } else {
                false
            }
        } catch (e: SQLException) {
            LogSmart.e("DBhelper", "SQLException в editDbUser($user)", e)
            false
        }
    }

    fun deleteDbUser(user: User): Boolean {
        // Функция удаляет контакт из БД.

        return try {
            writableDatabase.transaction {
                // Удаляем пользователя
                delete("users", "id = ?", arrayOf(user.id.toString()))
                // Удаляем его координаты
                delete("points", "phone = ?", arrayOf(user.phone))
            }
            // Если транзакция завершилась без ошибок — успех
            true
        } catch (e: SQLException) {
            LogSmart.e("DBhelper", "SQLException в deleteDbUser($user)", e)
            false
        }
    }

    fun writeDbLastPoint(record: PointRecord) {
        // Функция заносит в БД последние известные координаты контакта.

        if (record.latitude !in -90.0..90.0 || record.longitude !in -180.0..180.0) {
            return
        }

        val values = ContentValues().apply {
            put("phone", record.phone)
            put("latitude", String.format(Locale.US, PointRecord.FORMAT_DOUBLE, record.latitude))
            put("longitude", String.format(Locale.US, PointRecord.FORMAT_DOUBLE, record.longitude))
            put("datetime", record.dateTime)
        }

        try {
            writableDatabase.replace("points", null, values)
        } catch (e: SQLException) {
            LogSmart.e("DBhelper", "SQLException в writeDbLastPoint($record)", e)
        }
    }

    fun readDbLastAnswer(): PointRecord? {
        // Функция считывает из БД последнюю запись и возвращает PointRecord.
        // Возвращает null, если записи нет, или координаты выходят за допустимые границы.

        val query = "SELECT * FROM points ORDER BY datetime DESC LIMIT 1;"

        return readableDatabase.rawQuery(query, null).use { cursor ->
            val phoneIdx = cursor.getColumnIndexOrThrow("phone")
            val latIdx = cursor.getColumnIndexOrThrow("latitude")
            val lonIdx = cursor.getColumnIndexOrThrow("longitude")
            val dateIdx = cursor.getColumnIndexOrThrow("datetime")

            if (cursor.moveToFirst()) {
                try {
                    val record = PointRecord(
                        cursor.getString(phoneIdx),
                        cursor.getString(latIdx).toDouble(),
                        cursor.getString(lonIdx).toDouble(),
                        cursor.getString(dateIdx)
                    )

                    if (record.latitude !in -90.0..90.0 || record.longitude !in -180.0..180.0) {
                        null
                    } else { record }
                } catch (e: SQLException) {
                    LogSmart.e("DBhelper", "SQLException в readDbLastAnswer()", e)
                    null
                }
            } else { null }
        }
    }

    fun checkDbRecords() {
        // Функция удаляет из таблицы points записи, не имеющие ссылок на таблицу phones.
        // Это предотвращает накопление "мусорных" данных.

        try {
            writableDatabase.delete(
                "points",
                "phone NOT IN (SELECT phone FROM users)",
                null
            )
        } catch (e: SQLException) {
            LogSmart.e("DBhelper", "SQLException в checkDbRecords()", e)
        }
    }

}